corgidrp/flat.py

```diff
@@ -73,37 +73,93 @@ def raster_kernel(width, image, hard=True):
+#     matched_filters_smooth = [gauss(matched_filters[i],3) for i in range(len(matched_filters))] 
+    
+#     matched_residuals=[];
+#     matched_residuals_err=[]
+#     for j in range(len(raster_images)):
+#         n_images=int(np.floor(j//(len(raster_images)//len(matched_filters_smooth))))
+#         matched_residuals.append(raster_images[j]/matched_filters_smooth[n_images])
+#     return(matched_residuals)
+
 def flatfield_residuals(images, N=None):
     """Turn this dataset of image frames of neptune or uranus and create matched filters and estimate residuals after 
      dividing from matched filters
 
-     Todo: Propoagate the errors: incorporate the errors in the raster frames of neptune and jupiter and the errors in the matched filter
 
      Args:
     	images (np.array): 2D array of cropped neptune or uranus image frames
         N (int): number of images to be grouped. defaults to 3 for both Neptune and uranus. (If we use different number of dithers for neptune and uranus, option is provided to change N)
 
      Returns:
     	matched_residuals (np.array): residual image frames of neptune or uranus divided by matched filter
+        matched_residuals_err (np.array): erros of the residual image frames of neptune or uranus divided by matched filter
 	"""
     raster_images = np.array(images)
+    # M is the number of images in each group used to calculate a filter.
+    M = len(raster_images) // N
     images_split = np.array(np.split(np.array(raster_images),N))
     matched_filters = np.array([np.nanmedian(np.stack(images_split[i],2),2) for i in np.arange(0,len(images_split))])
     matched_filters_smooth = [gauss(matched_filters[i],3) for i in range(len(matched_filters))] 
     
+    #Estimate the initial uncertainty (standard error of the mean/median) for each filter group.
+    sigma_matched_filters = np.array([np.nanstd(images_split[i], axis=0) / np.sqrt(M) for i in np.arange(N)])
+    # Apply the same smoothing (Gauss(sigma=3)) to the error maps as was applied to the filters.
+    sigma_matched_filters_smooth = [gauss(sigma_matched_filters[i], 3) for i in range(N)]
+
+    
     matched_residuals=[];
     matched_residuals_err=[]
     for j in range(len(raster_images)):
+        
         n_images=int(np.floor(j//(len(raster_images)//len(matched_filters_smooth))))
-        matched_residuals.append(raster_images[j]/matched_filters_smooth[n_images])
-    return(matched_residuals)
+        current_image = raster_images[j]
+        
+        current_filter_smooth = matched_filters_smooth[n_images]
+        #calculate the residual
+        residual = current_image / current_filter_smooth
+        matched_residuals.append(residual)
+        
+        #error propagation for division (Z = X/Y) ---
+        # Sigma_Z = |Z| * sqrt( (Sigma_X/X)^2 + (Sigma_Y/Y)^2 )
+        
+        #sigma_x: Image error (Assumed Poisson noise: Sigma_X = sqrt(X))
+        sigma_image = np.sqrt(np.maximum(current_image, 1e-6))
```

Comment:
instead of just assuming poisson noise, why don't we just pass in the err maps of each frame as one of the inputs to the function?